File: iOS/Mobile/MenuTableViewHeaderFooterView.m
//
//  MenuTableViewHeaderFooterView.m
//  Mobile
//
//  Created by Jason Hocker on 4/4/14.
//  Copyright (c) 2014 Ellucian Company L.P. and its affiliates. All rights reserved.
//

#import "MenuTableViewHeaderFooterView.h"
#import "AppearanceChanger.h"

@implementation MenuTableViewHeaderFooterView

- (id)initWithReuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithReuseIdentifier:reuseIdentifier];
    if (self)
    {
        BOOL collapseable = [reuseIdentifier isEqualToString:@"CollapseableHeader"];
        UIColor *backgroundColor = [UIColor blackColor];
        if (![self.backgroundColor isEqual: backgroundColor]) {
            
            self.contentView.backgroundColor = backgroundColor;
            self.contentView.opaque = YES;
            
            UILabel* headerLabel = [UILabel new];
            _headerLabel = headerLabel;
            headerLabel.font = [UIFont boldSystemFontOfSize:14.0];
            headerLabel.textColor = [UIColor colorWithRed:179.0f/255.0f green:179.0f/255.0f blue:179.0f/255.0f alpha:1.0f];
            headerLabel.backgroundColor = [UIColor clearColor];
            headerLabel.opaque = NO;
            [self.contentView addSubview:headerLabel];
            headerLabel.translatesAutoresizingMaskIntoConstraints = NO;
            
            UIButton *collapsibleButton = nil;
            if(collapseable) {
                collapsibleButton = [UIButton new];
                _collapsibleButton = collapsibleButton;
                [collapsibleButton setImage:[UIImage imageNamed:@"menu header expanded"] forState:UIControlStateNormal];
                [collapsibleButton setImage:[UIImage imageNamed:@"menu header collapsed"] forState:UIControlStateSelected];
                [collapsibleButton addTarget:self
                                      action:@selector(toggleHeader:)forControlEvents:UIControlEventTouchUpInside];
                [self.contentView addSubview:collapsibleButton];
                collapsibleButton.translatesAutoresizingMaskIntoConstraints = NO;
            }
            
            NSDictionary *viewBindings = NSDictionaryOfVariableBindings(headerLabel);
            if([AppearanceChanger isRTL]) {
                if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
                    [self.contentView addConstraint:[NSLayoutConstraint constraintWithItem:headerLabel attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:self.contentView attribute:NSLayoutAttributeLeft multiplier:1.0 constant:307.0]];
                } else {
                    [self.contentView addConstraint:[NSLayoutConstraint constraintWithItem:headerLabel attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:self.contentView attribute:NSLayoutAttributeLeft multiplier:1.0 constant:267.0]];
                }
                if(collapseable) {
                    [self.contentView addConstraints:
                     [NSLayoutConstraint constraintsWithVisualFormat:@"H:[collapsibleButton]-13-|"
                                                             options:0 metrics:nil
                                                               views:@{@"collapsibleButton":collapsibleButton}]];
                }
            } else {
                [self.contentView addConstraints:
                 [NSLayoutConstraint constraintsWithVisualFormat:@"H:|-13-[headerLabel]"
                                                         options:0 metrics:nil
                                                           views:viewBindings]];
                if(collapseable) {
                    if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
                        [self.contentView addConstraint:[NSLayoutConstraint constraintWithItem:collapsibleButton attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:self.contentView attribute:NSLayoutAttributeLeft multiplier:1.0 constant:307.0]];
                    } else {
                        [self.contentView addConstraint:[NSLayoutConstraint constraintWithItem:collapsibleButton attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:self.contentView attribute:NSLayoutAttributeLeft multiplier:1.0 constant:267.0]];
                    }
                }
            }
            
            [self.contentView addConstraints:
             [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[headerLabel]|"
                                                     options:0 metrics:nil
                                                       views:viewBindings]];
            if(collapseable) {
                
                [self.contentView addConstraint: [NSLayoutConstraint constraintWithItem:collapsibleButton
                                                                              attribute:NSLayoutAttributeCenterY
                                                                              relatedBy:NSLayoutRelationEqual
                                                                                 toItem:headerLabel
                                                                              attribute:NSLayoutAttributeCenterY
                                                                             multiplier:1.0
                                                                               constant:0]];
            }
        }
        
        // set up the tap gesture recognizer
        if(collapseable) {
            UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self
                                                                                         action:@selector(toggleHeader:)];
            [self addGestureRecognizer:tapGesture];
        }
    }
    return self;
}

- (IBAction)toggleHeader:(id)sender {
    
    [self toggleHeaderWithUserAction:YES];
}

- (void)toggleHeaderWithUserAction:(BOOL)userAction {
    
    // toggle the disclosure button state
    self.collapsibleButton.selected = !self.collapsibleButton.selected;
    
    // if this was a user action, send the delegate the appropriate message
    if (userAction) {
        if (self.collapsibleButton.selected) {
            if ([self.delegate respondsToSelector:@selector(sectionHeaderView:sectionClosed:)]) {
                [self.delegate sectionHeaderView:self sectionClosed:self.section];
            }
        }
        else {
            if ([self.delegate respondsToSelector:@selector(sectionHeaderView:sectionOpened:)]) {
                [self.delegate sectionHeaderView:self sectionOpened:self.section];
            }
        }
    }
}

@end

File: iOS/Mobile/MenuTableViewHeaderFooterView.swift
//
//  MenuTableViewHeaderFooterView.swift
//  Mobile
//
//  Created by Jason Hocker on 7/13/15.
//  Copyright © 2015 Ellucian Company L.P. and its affiliates. All rights reserved.
//

import Foundation

class MenuTableViewHeaderFooterView : UITableViewHeaderFooterView {
    
    var headerLabel : UILabel?
    var collapsibleButton : UIButton?
    var delegate : SectionHeaderViewDelegate?
    var section : Int?
    
    override init(reuseIdentifier: String?) {
        super.init(reuseIdentifier: reuseIdentifier)
        
        let collapsible = reuseIdentifier == "CollapseableHeader"
        let backgroundColor = UIColor.blackColor()
        if self.backgroundColor != backgroundColor {
            self.contentView.backgroundColor = backgroundColor
            self.contentView.opaque = true
            
            headerLabel = UILabel()
            headerLabel!.font = UIFont.preferredFontForTextStyle(UIFontTextStyleHeadline)
            headerLabel!.textColor = UIColor(red: 179/255, green: 179/255, blue: 179/255, alpha: 1)
            headerLabel!.opaque = false
            self.contentView.addSubview(headerLabel!)
            headerLabel!.translatesAutoresizingMaskIntoConstraints = false
            headerLabel?.accessibilityTraits |= UIAccessibilityTraitHeader

    
            if(collapsible) {
                collapsibleButton = UIButton()
                collapsibleButton!.setImage(UIImage(named:"menu header expanded"), forState: .Normal)
                collapsibleButton!.setImage(UIImage(named:"menu header collapsed"), forState: .Selected)
                collapsibleButton!.addTarget(self, action: Selector("toggleHeader"), forControlEvents: .TouchUpInside)
                self.contentView.addSubview(collapsibleButton!)
                collapsibleButton!.translatesAutoresizingMaskIntoConstraints = false
                collapsibleButton!.accessibilityLabel = NSLocalizedString("Toggle menu section", comment:"Accessibility label for toggle menu section button")
            }

            if AppearanceChanger.isIOS8AndRTL() {
                self.contentView .addConstraint(NSLayoutConstraint(item: headerLabel!, attribute: .Right, relatedBy: .Equal, toItem: self.contentView, attribute: .Left, multiplier: 1.0, constant: 267))
                
                if(collapsible) {
                    self.contentView.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:[collapsibleButton]-13-|",
                        options:NSLayoutFormatOptions(rawValue: 0),
                        metrics: nil, views: ["collapsibleButton": collapsibleButton!]))
                }
            } else {
                self.contentView.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:|-13-[headerLabel]", options: NSLayoutFormatOptions(rawValue: 0), metrics: nil, views: ["headerLabel" : headerLabel!]))
                
                if(collapsible) {
                    self.contentView.addConstraint(NSLayoutConstraint(item: collapsibleButton!, attribute: .Trailing, relatedBy: .Equal, toItem: self.contentView, attribute: .Leading, multiplier: 1.0, constant: 267.0))
                }
            }
            
            self.contentView.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:|[headerLabel]|", options: NSLayoutFormatOptions(rawValue: 0), metrics: nil, views: ["headerLabel": headerLabel!]))
            
            if(collapsible) {
                self.contentView.addConstraint(NSLayoutConstraint(item: collapsibleButton!, attribute: .CenterY, relatedBy: .Equal, toItem: headerLabel!, attribute: .CenterY, multiplier: 1.0, constant: 0))
            }
        }
        
        if(collapsible) {
            let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: Selector("toggleHeader"))
            self.addGestureRecognizer(tapGestureRecognizer)
        }

    }
    
    required init?(coder aDecoder: NSCoder)
    {
        fatalError("init(coder:) has not been implemented")
    }
    
    func toggleHeader() {
        self.collapsibleButton!.selected = !self.collapsibleButton!.selected
        
        if self.collapsibleButton!.selected {
            self.delegate?.sectionHeaderView!(self, sectionClosed: self.section!)
        } else {
            self.delegate?.sectionHeaderView!(self, sectionOpened: self.section!)
        }
    }
}

